yo/app/scripts/directives/copy-or-migrate-mappings.js

```diff
@@ -0,0 +1,176 @@
+            ],
+            enableColumnResize: false,
+            selectedItems: $scope.selectedProject,
+            multiSelect: false
+        };
+
+        function createMigrationStatusMessage(inputDataModelId, referenceProjectId, type) {
+            var actionMsg = type;
+            var itemMsg = 'mappings from reference project ' + referenceProjectId + ' to a new project';
+            var additionalItemMsg = '';
+
+            if(inputDataModelId) {
+
+                additionalItemMsg = 'with help of the schema from input data model ' + inputDataModelId;
+            }
+
+            return function(statusMsg) {
+                return [actionMsg, itemMsg, additionalItemMsg, statusMsg].join(' ');
+            };
+        }
+
+        $scope.copyMappings = function() {
+
+            var selectedProject = $scope.selectedProject[0];
+            var referenceProjectId = selectedProject.uuid;
+
+            var payload = {
+                reference_project: referenceProjectId
+            };
+
+            var finishMessage = createMigrationStatusMessage(null, referenceProjectId, 'Copying');
+
+            ngProgress.start();
+
+            var dataModelIdPromise = $q.defer();
+
+
+
+            if($scope.selectedDataModel.length > 0) {
+
+                dataModelIdPromise.resolve($scope.selectedDataModel.uuid);
+            } else {
+
+                // utilise input data model from project
+                ProjectResource.get({id: referenceProjectId}, function(project) {
+
+                    dataModelIdPromise.resolve(project.input_data_model.uuid);
+                }, dataModelIdPromise.reject);
+            }
+
+            dataModelIdPromise.promise.then(function(dataModelId) {
+
+                payload.input_data_model = dataModelId;
+
+                finishMessage = createMigrationStatusMessage(dataModelId, referenceProjectId, 'Copying');
+
+                return ProjectResource.createProjectWithHelpOfExistingEntities({}, payload).$promise;
+            }).then(function() {
+
+
+                ngProgress.complete();
+                showAlert.show($scope, 'info', finishMessage('successfully finished.'));
+
+                $scope.updateGridData();
+
+                isNewProjectsHasBeenCreated = true;
+            }, function(resp) {
+
+                ngProgress.complete();
+                showAlert.show($scope, 'danger', finishMessage('failed.'), 5000);
+
+                console.log(resp);
+
+                $window.alert(resp.message || resp.data.error);
+            });
+        };
+
+        $scope.migrateMappings = function() {
+
+
+        };
+
+        $scope.showHelp = function() {
+
+            $modal.open({
+                templateUrl: 'views/controllers/copy-or-migrate-mappings-help.html'
+            });
+        };
+
+        $scope.save = function() {
+
+            var result = {
+                newProjectsHasBeenCreated: isNewProjectsHasBeenCreated
```

Comment:
yes, I'm not sure how stable this feature needs to be atm, i.e., I'm not sure how often it will be really applied in practice. So I'm not sure how stable it needs to be in its own ;)

I would follow "move fast, break things" here, i.e., we can make it stable later, or?
However, I you wish that it should be that stable right now, I can add the handling with the promises.